helix-lsp/src/snippet.rs

```diff
@@ -0,0 +1,525 @@
+    }
+
+    fn regex<'a>() -> impl Parser<'a, Output = Regex<'a>> {
+        let text = map(text(['$', '/']), FormatItem::Text);
+        let replacement = reparse_as(
+            take_until(|c| c == '/'),
+            one_or_more(choice!(format(), text)),
+        );
+
+        map(
+            seq!(
+                "/",
+                take_until(|c| c == '/'),
+                "/",
+                replacement,
+                "/",
+                optional(take_until(|c| c == '}')),
+            ),
+            |(_, value, _, replacement, _, options)| Regex {
+                value,
+                replacement,
+                options,
+            },
+        )
+    }
+
+    fn tabstop<'a>() -> impl Parser<'a, Output = SnippetElement<'a>> {
+        map(
+            or(
+                right("$", digit()),
+                map(seq!("${", digit(), "}"), |values| values.1),
+            ),
+            |digit| SnippetElement::Tabstop { tabstop: digit },
+        )
+    }
+
+    fn placeholder<'a>() -> impl Parser<'a, Output = SnippetElement<'a>> {
+        let text = map(text(['$', '}']), SnippetElement::Text);
+        map(
+            seq!(
+                "${",
+                digit(),
+                ":",
+                one_or_more(choice!(anything(), text)),
+                "}"
+            ),
+            |seq| SnippetElement::Placeholder {
+                tabstop: seq.1,
+                value: seq.3,
+            },
+        )
+    }
+
+    fn choice<'a>() -> impl Parser<'a, Output = SnippetElement<'a>> {
+        map(
+            seq!(
+                "${",
+                digit(),
+                "|",
+                sep(take_until(|c| c == ',' || c == '|'), ","),
+                "|}",
+            ),
+            |seq| SnippetElement::Choice {
+                tabstop: seq.1,
+                choices: seq.3,
+            },
+        )
+    }
+
+    fn variable<'a>() -> impl Parser<'a, Output = SnippetElement<'a>> {
+        choice!(
+            // $var
+            map(right("$", var()), |name| SnippetElement::Variable {
+                name,
+                default: None,
+                regex: None,
+            }),
+            // ${var:default}
+            map(
+                seq!("${", var(), ":", take_until(|c| c == '}'), "}",),
+                |values| SnippetElement::Variable {
+                    name: values.1,
+                    default: Some(values.3),
+                    regex: None,
+                }
+            ),
+            // ${var/value/format/options}
+            map(seq!("${", var(), regex(), "}"), |values| {
+                SnippetElement::Variable {
+                    name: values.1,
+                    default: None,
+                    regex: Some(values.2),
+                }
+            }),
+        )
+    }
+
+    fn anything<'a>() -> impl Parser<'a, Output = SnippetElement<'a>> {
+        // The parser has to be constructed lazily to avoid infinite recursion
```

Comment:
```suggestion
        // The parser has to be constructed lazily to avoid infinite opaque type recursion
```

Just to be clear that the recursion is a compile-time type checking issue rather than a potential runtime issue